Added ControllerRequestDelegateFactory

davidfowl opened the pull request:
- Today there's lots of indirection to figure out which invoker is responsible for executing a particular request based on the `ActionContext`. The `IActionInvokerFactory` calls through `IActionInvokerProviders` and an `IActionInvoker` is picked as a result of this. The default implementations based this decision on the type of action descriptor (though that's not required) which means its possible to avoid this per request indirection and select the invoker based on the descriptor. The `ControllerRequestDelegateFactory` creates a `RequestDelegate` for the endpoint don't go through indirection to find the ControllerActionInvoker.

**API**

```diff
public class MvcOptions
{
+   /// <summary>
+   /// Gets or sets the flag that determines if MVC should use action invoker extensibility. This will allow
+   /// custom <see cref="IActionInvokerFactory"/> and <see cref="IActionInvokerProvider"/> execute during the request pipeline.
+   /// </summary>
+   /// <remarks>This only applies when <see cref="EnableEndpointRouting"/> is true.</remarks>
+   public bool EnableActionInvokers { get; set; }
}
```

davidfowl committed:
Added IRequestDelegateFactory abstraction
- Today there's lots of indirection to figure out which invoker is
responsible for executing a particular request based on the `ActionContext`. The `IActionInvokerFactory` calls through
`IActionInvokerProviders` and an `IActionInvoker` is picked as a result of this. The default implementations based
this decision on the type of action descriptor (though that's not required) which means its possible to
avoid this per request indirection and select the invoker based on the descriptor.
The `ControllerRequestDelegateFactory` creates a `RequestDelegate` for the endpoint don't go
through indirection to find the ControllerActionInvoker.

PS: This code path breaks extensibility for IActionInvoker* types that want
to take over invocation of existing ControllerActionDescriptor since it no longer
calls them in the endpoint routing code path. This may warrant a compatibility switch

pranavkm commented on src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs:
```diff
@@ -273,6 +273,7 @@ internal static void AddMvcCoreServices(IServiceCollection services)
             services.TryAddSingleton<OrderedEndpointsSequenceProviderCache>();
             services.TryAddSingleton<ControllerActionEndpointDataSourceIdProvider>();
             services.TryAddSingleton<ActionEndpointFactory>();
+            services.TryAddSingleton<IRequestDelegateFactory, ControllerRequestDelegateFactory>();
```
Could you include the commit for `IRequestDelegateFactory` in this change? It seems sounds, but like you mentioned, we should put the use of this behind a compat flag.

davidfowl committed:
Added compatibility switch to enable action invokers
- This will allow people to turn off the request delegate factory if they were using that extensiblity point before.

rynowak commented:
> PS: This code path breaks extensibility for IActionInvoker* types that want to take over invocation of existing ControllerActionDescriptor since it no longer calls them in the endpoint routing code path. This may warrant a compatibility switch (this is why it's a draft)

FWIW - I'm not aware of anyone using this for extensiblity for anything real (along with filter providers). There were some projects using it back when our implementation of the invoker was public. Now that this requires some really complicated work, it's just not realistic for someone to extend.

rynowak commented on src/Mvc/Mvc.Core/src/Routing/ControllerRequestDelegateFactory.cs:
```diff
@@ -0,0 +1,101 @@
+
+                    controllerContext.ModelState.MaxAllowedErrors = _maxModelValidationErrors;
+
+                    var (cacheEntry, filters) = _controllerActionInvokerCache.GetCachedResult(controllerContext);
```
Another thing to investigate. This is a cache lookup that hits a concurrent dictionary (keyed on ActionDescriptor). Since you're creating a request delegate per-action there's no need for additional caches and all that stuff can get closed over here.

davidfowl replied:
https://github.com/dotnet/aspnetcore/pull/27763

rynowak commented on src/Mvc/Mvc.Core/src/Routing/ControllerRequestDelegateFactory.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class ControllerRequestDelegateFactory : IRequestDelegateFactory
```
Have you tried implementing this yet for Pages? I suspect it's more complex

davidfowl replied:
I have a stash for pages and it is more complex.

pranavkm commented on src/Mvc/Mvc.Core/src/MvcOptions.cs:
```diff
@@ -137,6 +138,13 @@ public MvcOptions()
+        /// Gets or sets the flag that determines if MVC should use action invoker extensibility. This will allow
+        /// custom <see cref="IActionInvokerFactory"/> and <see cref="IActionInvokerProvider"/> execute during the request pipeline.
+        /// </summary>
+        /// <remarks>This only applies when <see cref="EnableEndpointRouting"/> is true.</remarks>
```
```suggestion
        /// <remarks>This only applies when <see cref="EnableEndpointRouting"/> is true.</remarks>
        /// <value>Defaults to <see langword="false" />.</value>
```

pranavkm committed:
Merge branch 'master' into davidfowl/controller-delegate-factory

davidfowl committed:
Update src/Mvc/Mvc.Core/src/MvcOptions.cs

Co-authored-by: Pranav K <prkrishn@hotmail.com>

davidfowl committed:
Add public API

davidfowl commented:
/azp run

halter73 commented:
You said you were close to having an IRequestDelegateFactory implementation ready for Razor Pages. Can that be included in this PR?

halter73 reviewed (approved):
After some offline discussion, I'm fine with looking at the Razor Page change in a separate PR.

halter73 commented:
FWIW, here are the benchmark results for an early version of this PR hitting the [following action](https://github.com/aspnet/Benchmarks/blob/bb46db49c3b391291369d892719224f2f8a1fff8/scenarios/echo.yml) using [this crank config](https://github.com/aspnet/Benchmarks/blob/bb46db49c3b391291369d892719224f2f8a1fff8/scenarios/echo.yml):

```C#
    public class Person
    {
        public string Name { get; set; }
    }

    [ApiController]
    public class EchoController : ControllerBase
    {
        [HttpPost("/EchoController")]
        public Person Post(Person person) => person;
```

```
### MapController (master)

`> crank --config .\echo.yml --scenario mapcontroller --profile aspnet-perf-lin --application.options.outputFiles C:\dev\dotnet\aspnetcore\artifacts\bin\Microsoft.AspNetCore.Mvc.Core\Release\net6.0\`
| application           |                       |
| --------------------- | --------------------- |
| CPU Usage (%)         |                    98 |
| Raw CPU Usage (%)     |              1,176.68 |
| Working Set (MB)      |                   190 |
| Build Time (ms)       |                 3,614 |
| Start Time (ms)       |                   380 |
| Published Size (KB)   |                94,141 |
| .NET Core SDK Version | 5.0.100-rc.2.20479.15 |

| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          |        39 |
| Raw CPU Usage (%)      |    471.89 |
| Working Set (MB)       |        51 |
| Build Time (ms)        |     4,072 |
| Start Time (ms)        |       107 |
| Published Size (KB)    |    77,080 |
| .NET Core SDK Version  | 3.1.403   |
| First Request (ms)     |       189 |
| Requests               | 1,827,246 |
| Bad responses          |         0 |
| Mean latency (us)      |     2,096 |
| Max latency (us)       |    52,615 |
| Requests/sec           |   122,043 |
| Requests/sec (max)     |   132,081 |
| Read throughput (MB/s) |     20.56 |

### MapControler (davidfowl/fast-path)

`> crank --config .\echo.yml --scenario mapcontroller --profile aspnet-perf-lin --application.options.outputFiles C:\dev\dotnet\aspnetcore\artifacts\bin\Microsoft.AspNetCore.Mvc.Core\Release\net6.0\`

| application           |                       |
| --------------------- | --------------------- |
| CPU Usage (%)         |                    99 |
| Raw CPU Usage (%)     |              1,185.04 |
| Working Set (MB)      |                   186 |
| Build Time (ms)       |                 3,578 |
| Start Time (ms)       |                   372 |
| Published Size (KB)   |                94,141 |
| .NET Core SDK Version | 5.0.100-rc.2.20479.15 |

| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          |        42 |
| Raw CPU Usage (%)      |    507.19 |
| Working Set (MB)       |        51 |
| Build Time (ms)        |     4,105 |
| Start Time (ms)        |       106 |
| Published Size (KB)    |    77,080 |
| .NET Core SDK Version  | 3.1.403   |
| First Request (ms)     |       213 |
| Requests               | 2,061,249 |
| Bad responses          |         0 |
| Mean latency (us)      |     1,857 |
| Max latency (us)       |    63,930 |
| Requests/sec           |   137,556 |
| Requests/sec (max)     |   148,030 |
| Read throughput (MB/s) |     23.20 |
```

A ~7.5% RPS improvement for this action with a small JSON payload.

halter73 commented:
@davidfowl pointed out what I measured before removed a lot more stuff than this PR ultimately did. Here are the numbers for the actual PR:

```
### post-echo-controller (pre-PR)

| application           |          |
| --------------------- | -------- |
| CPU Usage (%)         | 99       |
| Raw CPU Usage (%)     | 1,186.47 |
| Working Set (MB)      | 187      |
| Build Time (ms)       | 3,596    |
| Start Time (ms)       | 383      |
| Published Size (KB)   | 94,119   |
| .NET Core SDK Version | 5.0.100  |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 39        |
| Raw CPU Usage (%)      | 468.89    |
| Working Set (MB)       | 51        |
| Build Time (ms)        | 4,118     |
| Start Time (ms)        | 108       |
| Published Size (KB)    | 77,077    |
| .NET Core SDK Version  | 3.1.404   |
| First Request (ms)     | 249       |
| Requests               | 1,845,112 |
| Bad responses          | 0         |
| Mean latency (us)      | 2,075     |
| Max latency (us)       | 79,749    |
| Requests/sec           | 123,451   |
| Requests/sec (max)     | 133,755   |
| Read throughput (MB/s) | 20.76     |

### post-echo-controller (post-PR)

| application           |          |
| --------------------- | -------- |
| CPU Usage (%)         | 99       |
| Raw CPU Usage (%)     | 1,182.20 |
| Working Set (MB)      | 188      |
| Build Time (ms)       | 3,617    |
| Start Time (ms)       | 384      |
| Published Size (KB)   | 94,119   |
| .NET Core SDK Version | 5.0.100  |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 39        |
| Raw CPU Usage (%)      | 472.26    |
| Working Set (MB)       | 50        |
| Build Time (ms)        | 4,104     |
| Start Time (ms)        | 106       |
| Published Size (KB)    | 77,077    |
| .NET Core SDK Version  | 3.1.404   |
| First Request (ms)     | 242       |
| Requests               | 1,873,259 |
| Bad responses          | 0         |
| Mean latency (us)      | 2,043     |
| Max latency (us)       | 62,319    |
| Requests/sec           | 125,211   |
| Requests/sec (max)     | 134,867   |
| Read throughput (MB/s) | 21.08     |
```

It looks like there might be a small perf improvement, but it's within the margin of error.
